raft-tech/TANF-app
# Issue 1260: Migrate Dev Deployments to Circle CI API V2

### Summary of Changes
Closes #1260 
Closes #1059 

* Updates the GitHub Action `deploy-on-label.yml` to utilize [a different open source action](https://github.com/promiseofcake/circleci-trigger-action) that uses the Circle CI V2 API and triggers a new pipeline with the provided parameters.
* Updates `.circleci/config.yml` to accept [Pipeline Parameters](https://circleci.com/docs/2.0/pipeline-variables/#pipeline-parameters-in-configuration) of `run_dev_deployment` and `target-env`.
* Updates `.circleci/config.yml` to utilize [Conditional Workflows](https://circleci.com/docs/2.0/pipeline-variables/#conditional-workflows) based on the provided parameters above. Specifically, the `build-and-test` workflow was updated to not run when `run_dev_deployment` is passed as true. A new workflow `dev-deployment` was added which will only run when the pipeline parameter `run_dev_deployment` is passed as true.

### How to Test
* Assign a `Deploy with CircleCI-{env}` label to this PR
* Ensure that the `dev-deployment` workflow runs independently from `build-and-test` in Circle CI
![Screenshot from 2021-08-30 17-02-22](https://user-images.githubusercontent.com/22626085/131405200-21c666b1-84fb-450a-b934-83cbf3ebe1fc.png)
* Ensure that both the `deploy-infrastructure-dev` (Terraform) and `deploy-dev` jobs are run
* Ensure that the appropriate `target-env` matching the label that was assigned is passed to the `deploy-dev` job as shown below:
![circleci-dev-deployment-v2](https://user-images.githubusercontent.com/22626085/131405053-5f0898bf-eb2d-4b12-890d-a3f17b86fa57.png)

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

  - [x] GitHub Action `deploy-on-label.yml` is updated to utilize the Circle CI V2 Pipelines API
  - [x] Workflows defined in `config.yml` can be run conditionally based on Pipeline Parameters

As facilitator/product manager, @kniz-raft will decide if ACs are met from Raft's perspective.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

+ Are all areas of code introduced in this PR meaningfully tested?
+ Are code coverage minimums met?

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*


### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*

    + [See the full Accessibility Assessment plan here.](https://github.com/HHS/TANF-app/blob/main/docs/a11y/how-18f-will-test-a11y.md)

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*


### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*


### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*
